### DISCUSSION TITLE
Estimating latent change score models in accelerated longitudinal designs with ctsem

### DISCUSSION THREAD

**Discussion by PFernandez-Cancer** [cdriveraus/ctsem#19] [Q&A]
Hi Charles, 

I'm trying to estimate a univariate LCS model with ctsem to analyse developmental data from an accelerated longitudinal design (Estrada and Ferrer, 2019). I'm using simulated data, so I know the true parameter values. My problem is that the estimates are not even close to their corresponding true values. I've estimated the same model with the multi-subject approximation from Hunter (2018) in OpenMx, and the parameter estimates were very good, so I guess I'm doing or interpreting something wrong. 

My data looks like this (10 out of 500 participants):

id cohort time age         Y
 1      1    1  33   20.115584
 1      1    2  69   22.012463
 1      1    3 127   24.953003
 1      1    4 169   27.919677
 2      1    1  23   14.784754
 2      1    2  94   16.750150
 2      1    3 133   22.084580
 2      1    4 165   20.963002
 3      1    1  10    3.141701
 3      1    2  86   12.493426

Just to clarify, the variable "age" represents the age in weeks, with age 0 being 5 years old (so for example, in row 2, age=69 is 5 years plus 69 weeks, and so on). This is the model I built:

LCS.ctsem <- ctModel(manifestNames = "Y", latentNames = c("y", "yA"),
                     DRIFT=c("beta",1,0,0),
                     LAMBDA = c(1, 0),
                     DIFFUSION = c(0,0,0,0), 
                     T0VAR = c("vary0", 0, "covy0yA", "varyA"),
                     T0MEANS = c("meany0", "meanyA"),
                     MANIFESTVAR = "mery",
                     MANIFESTMEANS = 0,
                     CINT = 0,
                     type="stanct", id="id", time="age")

where "yA" is the so-called additive component. I want to estimate the mean and variance of the initial latent state ("meany0" and "vary0"), mean and variance of the additive component ("meanyA" and "varyA"), the covariance between them ("covy0yA"), the autoregressive coefficient ("beta"), and the measurement error ("mery") (no dynamic error, so 7 parameters in total). When I fit the model, I obtain the following parameter estimates:

fit <- ctStanFit(long, ctstanmodel = LCS.ctsem)
summary(fit)

$popsd
         mean     sd   2.5%    50%  97.5%
meany0 7.8129 0.3164 7.1745 7.8069 8.4415
meanyA 2.8805 0.1021 2.6767 2.8795 3.0807

$popmeans
          mean     sd    2.5%     50%   97.5%
meany0 21.0639 0.3608 20.3415 21.0611 21.7666
meanyA 11.6194 0.0775 11.4697 11.6211 11.7664
beta   -0.4749 0.0042 -0.4835 -0.4748 -0.4666
mery    2.3020 0.0430  2.2189  2.3015  2.3860

And covy0yA = 22.5782

But the true parameter values (in continuous time) are very different:
beta = -.25
meany0 = 10
vary0 = 25
meanyA = 7.5
varyA = 1.5625
covy0yA = 4.375
merY = 2

I'm not sure what the problem is. I believe the model is correctly specified, and when I used OpenMx the parameter estimates were much closer to their true values. Also, I'm fitting the model to more simulated datasets using the same generating parameters, and the estimates are always similar to those above, so the problem is not related to that specific sample. My guess is that I have to apply some transformation to the parameters, but I'm not sure how. Do you see anything wrong in the code? Do I need to transform the parameters in order to interpret them?

Best regards,

Pablo F. Cáncer

References:
Estrada, E., & Ferrer, E. (2019). Studying developmental processes in accelerated cohort-sequential designs with discrete-and continuous-time latent change score models. Psychological methods, 24(6), 708.
Hunter, M. D. (2018). State space modeling in an open source, modular, structural equation modeling environment. Structural Equation Modeling: A Multidisciplinary Journal, 25(2), 307-324.

**Comment by cdriveraus** [comment]
Ok this is my 2nd response attempt, see below for first but I think I get it now -- you only have one observation per subject and you want the dynamics to capture change in the expectation across age? For that, you actually need to create a dummy 't0' observation for each subject where you want the t0 point to be. So, if you want t0means to reflect time 20, a) time 20 needs to be earlier or equal to the earliest observation in the data, and b) all subjects need an observation at time 20, which can very well be NA for all subjects. Otherwise, the t0means parameter gets applied to the first observation of each subject. It may be that this subtlety has not been communicated well in the manual. 

1st attempt, perhaps still informative:

Usually the only time transformations get confusing is on covariance parameters -- in this case because t0means is freely estimated, and is by default set to individually vary (indvarying = TRUE), your T0VAR specification for free t0means components (in this case all of them) is overridden. You can make things more closely matched by setting `LCS.ctsem$pars$indvarying <- FALSE` , but this is really just a cosmetic thing and shouldn't affect estimation (unless you haven't updated in some time and are suffering from an unfortunate bug in the individually varying covariance). 

All that aside, I don't see anything obviously wrong, but maybe it would help to include your data generation script, as the issue seems to be that there is some discrepancy between the generating model and the fitting model.

**Reply by PFernandez-Cancer** [reply]
Charles, thank you for your prompt response! The first column was actually the row names, sorry. I already edited that, as it was confusing. There are 4 observations per subject. 

Now that you mention it, I think I remember reading something in your articles or the manual about including a time 0, but I completely forgot about that. Anyway, you explained that perfectly in your response, and now I understand why the estimates were wrong. As you suggested, I included an age = 0 and Y = NA for every subject:

id cohort time age         Y
  1      1    1   0        NA
  1      1    2  10 18.314425
  1      1    3  88 23.841218
  1      1    4 138 28.078644
  1      1    5 162 29.152645
  2      1    1   0        NA
  2      1    2  33  8.853475

and it works perfectly. The parameter estimates are reasonable now. 

Thank you very much!!

**Reply by cdriveraus** [reply]
Aha good to hear :)

**Comment by PFernandez-Cancer** [comment]
Hi Charles, I have another question related to this work. I'm using the same dual LCS model, and I want to include cohort effects in the slope factor (or additive component) and in the self-feedback (i.e., autoregressive parameter). With the hierarchical approach available in ctsem, I can select the slope and the self-feedback as individual-varying parameters, and then include my cohort variable as a TIpredictor to account for the differences in these parameters across cohorts. 

However, in a LCS model, the self-feedback, and the slope are related. In particular, the mean asymptote of the trajectories is:
mean_asymptote = mean_slope / -self-feedback

And the variance in the asymptotes of the trajectories is:
var_asymptote = var_slope / (self-feedback)^2

In my simulation design, the asymptote of the trajectories does not change across cohorts, so when the self-feedback changes from one cohort to the next, the slope changes accordingly. Using the cohort as TIpredictor, I can estimate a different mean slope and self-feedback for each cohort, but I can't do the same with the **variance** of the slope, and the **covariance** between the initial state and the slope. One solution would be to allow these elements of the T0var matrix to be individual-varying and then use the cohort variable to predict them. I read in one of your papers thas this is problematic to estimate, and I see that ctsem fixes these parameters as individual-invariant if I try to do that.

My question is: Is there a way to estimate a different variance and covariance for each group/cohort of individuals? 
I'm going to think about other ways of including cohort effects in the variances, perhaps with phantom variables, but I'd like to know if there is a more straigthforward approach with ctsem.

**Reply by cdriveraus** [reply]
super quick response, tell me if you need more -- I may not have been specific enough writing about the t0variance because there are certainly cases where it can make sense, particularly with covariates as you are describing. In more vanilla models it doesn't make sense.

**Reply by PFernandez-Cancer** [reply]
I agree that it would make sense in this case, but when I fit the model (setting indvarying=TRUE for the mean, self-feedback, variance and covariance, and including the effect of the cohort in all these 4 parameters) I get this message during the estimation:

Maximum a posteriori estimation requested
Setting DRIFT diagonals of 0 to approximate 0 of -1e-06
**Individual variation in T0VAR parameters not possible, removed**
Using 2/12 logical CPU cores
Doing 1st pass optimize...

I'm not sure why it removes the individual variation in T0VAR, or what can I do to maintain individual variation there. In the summary output, it only estimates the cohort effect in the mean of the slope and the self-feedback, but not in the variance and covariance.

**Reply by cdriveraus** [reply]
I should make it possible to override that 'helper' of 'Individual variation in T0VAR parameters not possible, removed', and I should also find a way of making it clearer that fixed effects / covariates / time independent predictors *are* totally possible. So all you need to do is ignore / forget about the random effect on t0var and use a tipred for cohort.

**Reply by PFernandez-Cancer** [reply]
I think I understand now. I thought that, in order to estimate the cohort effect, "indvarying" had to be TRUE, but I see that I was wrong. Indvarying is used to estimate subject-level parameters, but the cohort acts like a "level 2" covariate. So ctsem estimates the effect of the cohort whether indvarying=TRUE or not, and after that, it calculates the subject level parameters (if indvarying=TRUE) or not (if indvarying=FALSE). Please correct me if I'm wrong.

However, I still have the same problem. ctsem estimates cohort effects in all the parameters of the model, **except for** the elements of the T0VAR matrix. But I need to estimate the effects on these parameters, because the variance of the slope is different for each cohort. If I ignore it, my results are biased. Let me show you my code in the attached file. I'll give you some simulated data too.

Perhaps there is something wrong with my specification of the T0VAR matrix, and that's the reason why ctsem doesn't estimate the cohort effect in these parameters. It does estimate it in all the other paramaters though, even in the manifest variance. 

[long_data.txt](https://github.com/cdriveraus/ctsem/files/7195715/long_data.txt)
[ctsemcode_LCS.txt](https://github.com/cdriveraus/ctsem/files/7195716/ctsemcode_LCS.txt)

**Reply by cdriveraus** [reply]
essentially 'indvarying' should be read as 'random individual variation', ie, not determined by covariates, which operate distinctly. Subject level parameters are available in either case. By default, covariates don't influence t0var, but you can turn this on by modifying the mymodel$pars tipred effect columns appropriately or by manually specifying the t0var matrix and using the | separator notation, so instead of for instance 
`T0VAR=matrix('t0var11', nrow=1, ncol=1)` 
you would have
`T0VAR=matrix('t0var11 |||| cohort', nrow=1, ncol=1)`

**Reply by PFernandez-Cancer** [reply]
I've done that. I've tried both modifying the mymodel$pars tipred effect and the manual specification with separators (although once you write TIpredNames=cohort in the model specification, all the cohort_effects are TRUE by default, so these modifications shouldn't be necessary). None of them works. It still doesn't estimate cohort effects in the variances/covariances of the T0VAR matrix. 

Here, in the summary output:
$tipreds
                     mean     sd    2.5%     50%   97.5%        z
tip_cohort_meany0 -1.5700 0.4078 -2.3487 -1.5566 -0.8508  -3.8500
tip_cohort_meanyA -0.7413 0.0654 -0.8707 -0.7402 -0.6139 -11.3393
tip_cohort_beta    0.0235 0.0042  0.0157  0.0232  0.0322   5.5372
tip_cohort_mery    0.0067 0.0073 -0.0077  0.0070  0.0206   0.9145

There should be additional rows for tip_cohort_varyA, tip_cohort_vary0, and tip_cohort_covy0yA.

**Reply by cdriveraus** [reply]
Ok. Sorry for confusion then. The issue is actually the message "Free T0VAR parameters as well as indvarying T0MEANS -- fixing T0VAR pars to diag matrix of 1e-3". basically, parameters for individual variation in t0means are not identified at the same time as the t0var parameters, and ctsem makes the assumption that you want the indvarying parameters because they also covary with other individually varying parameters. In this case however, you don't have any other indvarying parameters and you actually just need to disable indvarying on the t0means. This is an awkward and unintuitive situation I realise, I'll have a think about how to improve it but if you've any suggestions for what would have made it clearer for you I'd be happy to hear! :) 

```
library(ctsem)
long <- read.table("long_data.txt")

LCS.ctsem <- ctModel(manifestNames = "Y", latentNames = c("y", "yA"),
  TIpredNames = "cohort",
  DRIFT=c("beta",1,0,0),
  LAMBDA = c(1, 0),
  DIFFUSION = c(.0,0,0,0), 
  T0VAR = c("vary0||||cohort", 0, 
    "covy0yA||||cohort", "varyA||||cohort"),
  T0MEANS = c("meany0||FALSE", "meanyA||FALSE"), #disabled indvarying
  MANIFESTVAR = "mery",
  MANIFESTMEANS = 0,
  type="stanct", id="id", time="age")

LCS.ctsem$pars$indvarying <- FALSE #alternative way, disables all indvarying

fit <- ctStanFit(long, ctstanmodel = LCS.ctsem, noprior=F)

s <- summary(fit)
print(s)
```

**Reply by PFernandez-Cancer** [reply]
Thank you very much Charles, it works now. Actually, I don't know why, but I accidentally didn't paste that warning message in my previous comment, so your confusion was completely understandable. 

Your explanation makes a lot of sense now that I think about it. I think most of my confusion stemmed from my misconception that individual variation in the parameters was a pre-requisite for estimating tipred effects, but you already clarified that it doesn't work that way.  Actually, there are situations like this, where the individual variation in the means is not compatible with tipred effects on the variances. If you ask me for suggestions, perhaps simply clarifying these ideas in a few lines in the manual will be helpful for some.
Thank you again! :)

**Comment by PFernandez-Cancer** [comment]
Charles, I'm afraid I'm back again with more questions :)

I've been comparing OpenMx and ctsem with the same model and data I sent you the last time, with cohort effects only in the self-feedback (beta), mean and variance of the slope (meanyA and varyA), and covariance (covy0yA). In OpenMx, I used mxAlgebra() to include cohort effects. The generating parameters were (very) accurately recovered. But the same model in ctsem leads to bias in some parameters. I'm wondering if the operations used by ctsem to include tipred effects are different from the matrix algebra that I used in OpenMx. For example:

The beta, yAmn, and y0Acv in my mxAlgebra are defined as:
beta = beta00 + lambda_beta * data.cohort
meanyA = meanyA00 + lambda_yAmn * data.cohort
covy0yA = covy0yA00 + lambda_y0Acv * data.cohort

where beta00, meanyA00, covy0yA00 and the lambdas are freely estimated. If I'm not mistaken, ctsem should perform these same operations (or similar), and output the estimated values of the lambdas (in the $tipreds table of the summary output).

In the case of varyA, the equation is a bit more complicated, but necessary to properly account for the effect of the cohort in the variance:
varyA = sdyA^2 + (lambda_yAv * data.coh)^2 + 2 * sdyA * lambda_yAv
where sdyA and lambda_yAv are freely estimated (and sdyA is the standard deviation of the slope yA).

In ctsem all these lambdas (I mean, the tipred effects) are more biased than I'd expect, especially the tipred effect on the covariance (lambda_y0Acv?).

My questions are: 1) Is ctsem using these same operations to include tipred effects (i.e. are my lambdas and your tipred effects the same thing)? 2) Do you know of any reason why lambda_y0Acv may be so biased? Maybe it needs to be transformed/rescaled somehow?

**Reply by cdriveraus** [reply]
quick answer, tell me if i'm off track -- you can't effectively parameterise a covariance parameter like that because it will lead to non positive definite matrices in higher dimensions, so ctsem works with a transformation of correlation parameters in the off-diagonals as a basis. When system matrices are reported, these are covariance matrices, but when the parameters and regression beta's are reported, these refer to the unconstrained correlations. It will require some fiddling around to get the equivalent parameters, but it is do-able using the extracted samples from ctExtract and computing the appropriate transformations of the matrices. Not sure how critical it is for you.
